manual/src/tutorials/memorymodel.etex

```diff
@@ -0,0 +1,790 @@
+between domains.
+
+\begin{caml_example*}{verbatim}
+(* No data race *)
+let msg = ref 0
+let flag = Atomic.make false
+let d1 () =
+  msg := 42; (* a *)
+  Atomic.set flag true (* b *)
+let d2 () =
+  if Atomic.get flag (* c *) then
+    !msg (* d *)
+  else 0
+\end{caml_example*}
+
+Observe that the actions "a" and "d" write and read from the same non-atomic
+location "msg", respectively, and hence are conflicting. We need to establish
+that "a" and "d" have a happens-before relationship in order to show that this
+program does not have a data race.
+
+The action "a" precedes "b" in program order, and hence, "a" happens-before
+"b". Similarly, "c" happens-before "d". If "d2" observes the atomic variable
+"flag" to be "true", then "b" precedes "c" in happens-before order. Since
+happens-before is transitive, the conflicting actions "a" and "d" are in
+happens-before order. If "d2" observes the "flag" to be "false", then the read
+of "msg" is not done. Hence, there is no conflicting access in this execution
+trace. Hence, the program does not have a data race.
+
+The following modified version of the message passing program does have a data
+race.
+
+\begin{caml_example*}{verbatim}
+(* Has data race *)
+let msg = ref 0
+let flag = Atomic.make false
+let d1 () =
+  msg := 42; (* a *)
+  Atomic.set flag true (* b *)
+let d2 () =
+  ignore (Atomic.get flag); (* c *)
+  !msg (* d *)
+\end{caml_example*}
+
+The domain "d2" now unconditionally reads the non-atomic reference "msg".
+Consider the execution trace:
+
+\begin{verbatim}
+Atomic.get flag; (* c *)
+!msg; (* d *)
+msg := 42; (* a *)
+Atomic.set flag true (* b *)
+\end{verbatim}
+
+In this trace, "d" and "a" are conflicting operations. But there is no
+happens-before relationship between them. Hence, this program has a data race.
+
+\section{s:local_drf}{Local data race freedom}
+
+The OCaml memory model offers strong guarantees even for programs with data races.
+It offers what is known as \emph{local data race freedom sequential
+consistency (LDRF-SC)} guarantee. A formal definition of this property is beyond
+the scope of this manual chapter. Interested readers are encouraged to read the
+PLDI 2018 paper on \href{https://doi.org/10.1145/3192366.3192421}{Bounding Data
+Races in Space and Time}.
+
+Informally, LDRF-SC says that the data race free parts of the program remain
+sequentially consistent. That is, even if the program has data races, those
+parts of the program that are disjoint from the parts with data races are
+amenable to sequential reasoning.
+
+Consider the following snippet:
+
+\begin{caml_example*}{verbatim}
+let snippet () =
+  let c = ref 0 in
+  c := 42;
+  let a = !c in
+  a
+\end{caml_example*}
+
+Observe that "c" is a newly allocated reference. Can the read of "c" return a
+value which is not 42? Surprisingly, in the C++ and Java memory models, the
+answer is yes. With the C++ memory model, if the program has a data race, even
```

Comment:
Well, this is not true in Java if `snippet` does not expose `c` to the rest of the world. So we should make `snippet` return `c` as well.

Reply:
How do you want this to be done? Could you suggest the edit here?